bin/migrate-wallets-to-grants.js

```diff
@@ -0,0 +1,156 @@
+on conflict (id)
+do update
+set
+    provider = $2,
+    provider_id = $3,
+    public_key = $4,
+    anonymous_address = $5,
+    provider_linking_id = $6;
+`
+
+async function main () {
+  const {
+    START_TIME: startTime = new Date(),
+    MONGODB_URI: mongodbURI,
+    DATABASE_URL: postgresURL
+  } = process.env
+  const postgres = new Postgres({ postgres: { url: postgresURL, roURL: false } })
+  const { mongo: db, mongoClient } = await mongo({ url: mongodbURI })
+  // wallets
+
+  let counter = 0
+  let complete = 0
+  try {
+    const wallets = db.collection('wallets')
+    const maxObjectID = objectIdFromTimestamp(startTime)
+    const query = {
+      _id: {
+        $lte: maxObjectID
+      }
+    }
+    console.log('document count', await wallets.countDocuments(query))
+    const batchSize = 1000
+    const cursor = await wallets.find(query, {
+      batchSize
+    })
+    const insertions = []
+    await postgres.connect()
+    const id = setInterval(() => (
+      console.log('progress', counter, complete)
+    ), 10000)
+    let shouldContinue = false
+    do {
+      console.log('waiting for', insertions.length)
+      await Promise.all(insertions)
+      console.log('insertions.length', insertions.length)
+      if (insertions.length) {
+        throw new Error('too many to finish')
+      }
+      while (await cursor.hasNext()) {
+        const wallet = await cursor.next()
+        if (wallet.httpSigningPubKey) {
+          const insert = postgres.query(insertStatement, [
+            wallet.paymentId,
+            wallet.provider,
+            wallet.providerId,
+            wallet.httpSigningPubKey,
+            wallet.anonymousAddress || null,
+            wallet.providerLinkingId || null
+          ]).then(() => {
+            insertions.splice(insertions.indexOf(insert), 1)
+            complete += 1
+          }).catch((err) => {
+            console.log('erred', err)
+            insertions.splice(insertions.indexOf(insert), 1)
+            complete += 1
+          })
+          counter += 1
+          insertions.push(insert)
+          if (insertions.length >= batchSize) {
+            break
+          }
+        } else {
+          console.log('erred', wallet.paymentId)
+        }
+      }
+      shouldContinue = insertions.length
+    } while (shouldContinue)
+    clearTimeout(id)
+  } catch (err) {
+    console.log(counter, complete)
+    console.log(err)
+  } finally {
+    await mongoClient.close()
+  }
+}
+
+main().then(result => {}).catch(e => {
+  console.error(e)
+})
+
+function objectIdFromTimestamp (_timestamp) {
+  // Convert string date to Date object (otherwise assume timestamp is a date)
+  const timestamp = new Date(_timestamp)
+  // Convert date object to hex seconds since Unix epoch
+  var hexSeconds = Math.floor(timestamp / 1000).toString(16)
+  // Create an ObjectId with that hex timestamp
+  return bson.ObjectID.createFromHexString(hexSeconds + '0000000000000000')
+}
+
+function pathname (url) {
```

Comment:
can you just construct a node URL object and get the pathname there? https://nodejs.org/api/url.html#url_url_pathname

Reply:
unfortunatley, i tried this, and the reason i went around was to because whenever urls have a shard as many mongo ones on heroku do, an `ERR_INVALID_URL` error is sent back when parsing the url. however, there is the parse member which can be used for this purpose with an eslint disable comment as done: 
bf49d4d